src/text-editor.coffee

```diff
@@ -818,114 +818,133 @@ class TextEditor extends Model
+          selectionsToMove.push(selection)
+          rangesOfSelectionsToMove.push(lastSelectionRange)
+
+        # Compute the range spanned by all these selections...
+        linesRangeStart = [selectionsToMove[0].getBufferRange().start.row, 0]
+        if lastSelectionRange.end.row > lastSelectionRange.start.row and lastSelectionRange.end.column is 0
+          # Don't move the last line of a multi-line selection if the selection ends at column 0
+          linesRange = new Range(linesRangeStart, lastSelectionRange.end)
         else
-          startRow = row
-          endRow = row
+          linesRange = new Range(linesRangeStart, [lastSelectionRange.end.row + 1, 0])
 
-        insertPosition = Point.fromObject([startRow - insertDelta])
-        endPosition = Point.min([endRow + 1], @buffer.getEndPosition())
-        lines = @buffer.getTextInRange([[startRow], endPosition])
-        if endPosition.row is lastRow and endPosition.column > 0 and not @buffer.lineEndingForRow(endPosition.row)
-          lines = "#{lines}\n"
+        # If selected line range is preceded by a fold, one line above on screen
+        # could be multiple lines in the buffer.
+        precedingScreenRow = @screenRowForBufferRow(linesRange.start.row) - 1
+        precedingBufferRow = @bufferRowForScreenRow(precedingScreenRow)
+        insertDelta = linesRange.start.row - precedingBufferRow
 
-        @buffer.deleteRows(startRow, endRow)
+        # Any folds in the text that is moved will need to be re-created.
+        rangesToRefold =
+          @outermostFoldsInBufferRowRange(linesRange.start.row, linesRange.end.row).map (fold) ->
+            fold.getBufferRange().translate([-insertDelta, 0])
 
         # Make sure the inserted text doesn't go into an existing fold
-        if fold = @displayBuffer.largestFoldStartingAtBufferRow(insertPosition.row)
-          @unfoldBufferRow(insertPosition.row)
-          foldedRows.push(insertPosition.row + endRow - startRow + fold.getBufferRange().getRowCount())
+        if fold = @displayBuffer.largestFoldStartingAtBufferRow(precedingBufferRow)
+          rangesToRefold.push(fold.getBufferRange().translate([linesRange.getRowCount(), 0]))
+          fold.destroy()
 
-        @buffer.insert(insertPosition, lines)
+        # Delete lines spanned by selection and insert them on the preceding buffer row
+        lines = @buffer.getTextInRange(linesRange)
+        lines += @buffer.lineEndingForRow(linesRange.end.row - 1) unless lines[lines.length - 1] is '\n'
+        @buffer.delete(linesRange)
+        @buffer.insert([precedingBufferRow, 0], lines)
 
-      # Restore folds that existed before the lines were moved
-      for foldedRow in foldedRows when 0 <= foldedRow <= @getLastBufferRow()
-        @foldBufferRow(foldedRow)
+        # Restore folds that existed before the lines were moved
+        for rangeToRefold in rangesToRefold
+          @displayBuffer.createFold(rangeToRefold.start.row, rangeToRefold.end.row)
```

Comment:
@lpommers Definitely looks wrong, but I think we should correct the problem when we're originally pushing the range onto `rangesToRefold`. Would be good to add a spec that reproduces the behavior you've pictured.